src/biokbase/narrative/tests/generate_test_results.py

```diff
@@ -42,10 +37,37 @@ def get_test_spec(tag, app_id):
     return copy.deepcopy(TEST_SPECS[tag][app_id])
 
 
+def generate_mappings(all_jobs):
+    # collect retried jobs and generate the cell-to-job mapping
+    retried_jobs = {}
+    jobs_by_cell_id = {}
+    batch_jobs = set()
+    for job in all_jobs.values():
+        if "batch_job" in job and job["batch_job"]:
+            batch_jobs.add(job["job_id"])
+        # save the first retry ID with the retried job
+        if "retry_ids" in job and len(job["retry_ids"]) > 0:
+            retried_jobs[job["job_id"]] = job["retry_ids"][0]
+
+        cell_id = job.get("job_input", {}).get("narrative_cell_info", {}).get("cell_id", None)
+
+        # add the new job to the jobs_by_cell_id mapping if there is a cell_id present
+        if cell_id:
+            if cell_id not in jobs_by_cell_id.keys():
+                jobs_by_cell_id[cell_id] = set()
+
+            jobs_by_cell_id[cell_id].add(job["job_id"])
+            if job["batch_id"]:
+                jobs_by_cell_id[cell_id].add(job["batch_id"])
+
+    return (retried_jobs, jobs_by_cell_id, batch_jobs)
+
+
 def _generate_job_output(job_id):
     state = get_test_job(job_id)
     job_input = state.get("job_input", {})
     narr_cell_info = job_input.get("narrative_cell_info", {})
+    widget_info = state.get("widget_info", None)
```

Comment:
Slight alteration to the way that the tests deal with widget info: in reality widget info is generated dynamically by the Job module, but rather than have this test data rely on an existing Job method, the widget data is now included as part of the ee2 test data as an extra 'widget_info' field.